File: C#/Question2_Solution2/NumberFinder.cs
using System;
using System.Collections.Generic;

namespace Question2_Solution2
{
    public class NumberFinder
    {
        private readonly List<NumberClass> numbersToCheck = new List<NumberClass>();

        public NumberFinder()
        {
            numbersToCheck.Add(new NumberClass(1));
        }


        public long ReachTargetNumber(int time)
        {
            long result = 1;
            
            for (long i = 0; i < time-1; i++)
            {
                result = GetNextSmallestInheritor();
            }

            return result;

        }

        private long GetNextSmallestInheritor()
        {
            long smallestInheritor = -1;
            bool hasInitialized = false;
            List<NumberClass> numbersToMovePointer = new List<NumberClass>();
            foreach (NumberClass numberToCheck in numbersToCheck)
            {
                if (!hasInitialized)
                {
                    smallestInheritor = numberToCheck.SmallestInheritor();
                    numbersToMovePointer.Add(numberToCheck);
                    hasInitialized = true;
                }
                else
                {
                    if (numberToCheck.SmallestInheritor() < smallestInheritor)
                    {
                        smallestInheritor = numberToCheck.SmallestInheritor();
                        numbersToMovePointer.Clear();
                        numbersToMovePointer.Add(numberToCheck);
                    }
                    else if (numberToCheck.SmallestInheritor() == smallestInheritor)
                    {
                        smallestInheritor = numberToCheck.SmallestInheritor();
                        numbersToMovePointer.Add(numberToCheck);
                    }
                }
            }


            foreach (NumberClass numberClass in numbersToMovePointer)
            {
                if (!numberClass.MovePointer())
                {
                    numbersToCheck.Remove(numberClass);
                }
            }


            bool shouldCreateNewNumber = true;
            foreach (NumberClass numberClass in numbersToCheck)
            {
                if (numberClass.TargetNumber == smallestInheritor)
                {
                    shouldCreateNewNumber = false;
                }
            }

            if (shouldCreateNewNumber)
            {
                CreateNewNumber(smallestInheritor);
            }

            return smallestInheritor;
            
            
        }

        private void CreateNewNumber(long number)
        {
            numbersToCheck.Add(new NumberClass(number));
        }
    }
}

File: C++/Question2_Solution2/NumberFinder.cpp
//
// Created by galax on 3/19/2020.
//

#include <iostream>
#include "NumberFinder.h"

long NumberFinder::ReachTargetNumber(int time) {
    long result = 1;
    for (long i = 0; i < time - 1; i++) {
        result = GetNextSmallestInheritor();
    }
    return result;
}

long NumberFinder::GetNextSmallestInheritor() {
    bool hasInitialized = false;

    long smallestInheritor = -1; // the smallest number that satisfy the condition yet bigger than all the numbers already in the target list
    vector<NumberClass*> numbersToMovePointer;   // The pointers of these number class need to be move forward by one

    for (auto numberToCheck: numbersToCheck) {
        if (!hasInitialized) {
            smallestInheritor = numberToCheck->SmallestCombination();
            numbersToMovePointer.push_back(numberToCheck);
            hasInitialized = true;
        } else {
            if (numberToCheck->SmallestCombination() < smallestInheritor) {
                smallestInheritor = numberToCheck->SmallestCombination();
                numbersToMovePointer.clear();
                numbersToMovePointer.push_back(numberToCheck);


            } else if (numberToCheck->SmallestCombination() == smallestInheritor) {
                smallestInheritor = numberToCheck->SmallestCombination();
                numbersToMovePointer.push_back(numberToCheck);
            }
        }
    }


    for (int i = 0; i < numbersToMovePointer.size(); ++i) {
        if (!numbersToMovePointer[i]->MovePointer()) {
            numbersToCheck.erase(numbersToCheck.begin() + i);
        }
    }

    bool shouldCreateNewNumber = true;
    for (auto numberClass: numbersToCheck) {
        if (numberClass->targetNumber == smallestInheritor) {
            shouldCreateNewNumber = false;
        }
    }
    if (shouldCreateNewNumber) {
        CreateNewNumber(smallestInheritor);
    }
    return smallestInheritor;

}

void NumberFinder::CreateNewNumber(long number) {
    numbersToCheck.push_back(new NumberClass(number));
}

NumberFinder::NumberFinder() {
    numbersToCheck.push_back(new NumberClass(1));
}
